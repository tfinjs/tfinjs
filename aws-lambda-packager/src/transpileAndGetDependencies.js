import extractDependenciesFromWebpackStats from './extractDependenciesFromWebpackStats';
import Compiler from './Compiler';
import processAndFilterDependencies from './processAndFilterDependencies';

const transpileAndGetDependencies = async (entryFilePath) => {
  const compilerWithDeps = new Compiler('analyzeDeps', {
    entry: entryFilePath,
    withNodeExternals: false,
  });

  const compiler = new Compiler('noDeps', {
    entry: entryFilePath,
    withNodeExternals: true,
  });

  const [noDeps, withDeps] = await Promise.all([
    compiler.run(),
    compilerWithDeps.run(),
  ]);

  const dependencyNames = extractDependenciesFromWebpackStats(withDeps.stats);

  const dependencies = processAndFilterDependencies({
    dependencyNames,
    entryFilePath,
  });

  return {
    data: noDeps.fs.data[Object.keys(noDeps.fs.data)[0]],
    dependencies,
  };
};

export default transpileAndGetDependencies;
