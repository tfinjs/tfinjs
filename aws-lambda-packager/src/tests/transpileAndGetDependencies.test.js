/* eslint-env jest */
import { resolve, relative } from 'path';
import transpileAndGetDependencies from '../transpileAndGetDependencies';

test('transpileAndGetDependencies', async () => {
  const entryFilePath = resolve(__dirname, 'src/service.js');
  const { dependencies } = await transpileAndGetDependencies(entryFilePath);
  const projectRootFolderPath = resolve(__dirname, '../../../');
  const normalizedDeps = dependencies.map((ob) => ({
    ...ob,
    ...[
      'dependencyFolderPath',
      'yarnLockFilePath',
      'packageJsonFilePath',
      'projectFolder',
    ].reduce(
      (c, key) => ({
        ...c,
        [key]: relative(projectRootFolderPath, ob[key]),
      }),
      {},
    ),
  }));
  expect(normalizedDeps).toMatchSnapshot();
});
