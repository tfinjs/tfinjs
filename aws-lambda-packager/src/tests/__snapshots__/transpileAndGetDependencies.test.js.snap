// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transpileAndGetDependencies 1`] = `
Array [
  Object {
    "dependencyFolderPath": "aws-lambda-packager/src/tests/src/node_modules/lodash",
    "dependencyName": "lodash",
    "packageJsonFilePath": "aws-lambda-packager/src/tests/src/package.json",
    "projectFolder": "aws-lambda-packager/src/tests/src",
    "yarnLockFilePath": "aws-lambda-packager/src/tests/src/yarn.lock",
  },
  Object {
    "dependencyFolderPath": "aws-lambda-packager/node_modules/source-map",
    "dependencyName": "source-map",
    "packageJsonFilePath": "aws-lambda-packager/package.json",
    "projectFolder": "aws-lambda-packager",
    "yarnLockFilePath": "aws-lambda-packager/yarn.lock",
  },
  Object {
    "dependencyFolderPath": "aws-lambda-packager/node_modules/debug",
    "dependencyName": "debug",
    "packageJsonFilePath": "aws-lambda-packager/package.json",
    "projectFolder": "aws-lambda-packager",
    "yarnLockFilePath": "aws-lambda-packager/yarn.lock",
  },
  Object {
    "dependencyFolderPath": "aws-lambda-packager/node_modules/resolve",
    "dependencyName": "resolve",
    "packageJsonFilePath": "aws-lambda-packager/package.json",
    "projectFolder": "aws-lambda-packager",
    "yarnLockFilePath": "aws-lambda-packager/yarn.lock",
  },
  Object {
    "dependencyFolderPath": "node_modules/supports-color",
    "dependencyName": "supports-color",
    "packageJsonFilePath": "package.json",
    "projectFolder": "",
    "yarnLockFilePath": "yarn.lock",
  },
  Object {
    "dependencyFolderPath": "node_modules/has-flag",
    "dependencyName": "has-flag",
    "packageJsonFilePath": "package.json",
    "projectFolder": "",
    "yarnLockFilePath": "yarn.lock",
  },
  Object {
    "dependencyFolderPath": "node_modules/color-convert",
    "dependencyName": "color-convert",
    "packageJsonFilePath": "package.json",
    "projectFolder": "",
    "yarnLockFilePath": "yarn.lock",
  },
  Object {
    "dependencyFolderPath": "aws-lambda-packager/node_modules/json5",
    "dependencyName": "json5",
    "packageJsonFilePath": "aws-lambda-packager/package.json",
    "projectFolder": "aws-lambda-packager",
    "yarnLockFilePath": "aws-lambda-packager/yarn.lock",
  },
  Object {
    "dependencyFolderPath": "aws-lambda-packager/src/tests/src/node_modules/@sindresorhus/is",
    "dependencyName": "@sindresorhus/is",
    "packageJsonFilePath": "aws-lambda-packager/src/tests/src/package.json",
    "projectFolder": "aws-lambda-packager/src/tests/src",
    "yarnLockFilePath": "aws-lambda-packager/src/tests/src/yarn.lock",
  },
  Object {
    "dependencyFolderPath": "aws-lambda-packager/node_modules/ms",
    "dependencyName": "ms",
    "packageJsonFilePath": "aws-lambda-packager/package.json",
    "projectFolder": "aws-lambda-packager",
    "yarnLockFilePath": "aws-lambda-packager/yarn.lock",
  },
  Object {
    "dependencyFolderPath": "node_modules/chalk",
    "dependencyName": "chalk",
    "packageJsonFilePath": "package.json",
    "projectFolder": "",
    "yarnLockFilePath": "yarn.lock",
  },
  Object {
    "dependencyFolderPath": "node_modules/escape-string-regexp",
    "dependencyName": "escape-string-regexp",
    "packageJsonFilePath": "package.json",
    "projectFolder": "",
    "yarnLockFilePath": "yarn.lock",
  },
  Object {
    "dependencyFolderPath": "node_modules/ansi-styles",
    "dependencyName": "ansi-styles",
    "packageJsonFilePath": "package.json",
    "projectFolder": "",
    "yarnLockFilePath": "yarn.lock",
  },
  Object {
    "dependencyFolderPath": "node_modules/color-name",
    "dependencyName": "color-name",
    "packageJsonFilePath": "package.json",
    "projectFolder": "",
    "yarnLockFilePath": "yarn.lock",
  },
  Object {
    "dependencyFolderPath": "aws-lambda-packager/node_modules/path-parse",
    "dependencyName": "path-parse",
    "packageJsonFilePath": "aws-lambda-packager/package.json",
    "projectFolder": "aws-lambda-packager",
    "yarnLockFilePath": "aws-lambda-packager/yarn.lock",
  },
  Object {
    "dependencyFolderPath": "aws-lambda-packager/node_modules/safe-buffer",
    "dependencyName": "safe-buffer",
    "packageJsonFilePath": "aws-lambda-packager/package.json",
    "projectFolder": "aws-lambda-packager",
    "yarnLockFilePath": "aws-lambda-packager/yarn.lock",
  },
]
`;
